dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp

```diff
@@ -210,6 +140,412 @@ void DeltaMergeStore::ingestFiles(
+         *    │--------------- Segment ---------------│
+         *             │-------- Ingest Range --------│
+         *
+         * We will logical split the segment to form a perfect matching segment:
+         *    │--------│------ Segment ---------------│
+         *             │-------- Ingest Range --------│
+         */
+        const auto [left, right] = segmentSplit(dm_context, segment, SegmentSplitReason::IngestBySplit, ingest_range.start, SegmentSplitMode::Logical);
+        UNUSED(left);
+        UNUSED(right);
+        return false;
+    }
+    else
+    {
+        /*
+         * Example:
+         *    │--------------- Segment ---------------│
+         *        │-------- Ingest Range --------│
+         *
+         * We invoke a logical split first:
+         *    │---│----------- Segment ---------------│
+         *        │-------- Ingest Range --------│
+         */
+        const auto [left, right] = segmentSplit(dm_context, segment, SegmentSplitReason::IngestBySplit, ingest_range.start, SegmentSplitMode::Logical);
+        UNUSED(left);
+        UNUSED(right);
+        return false;
+    }
+}
+
+void DeltaMergeStore::ingestFiles(
+    const DMContextPtr & dm_context,
+    const RowKeyRange & range,
+    const std::vector<DM::ExternalDTFileInfo> & external_files,
+    bool clear_data_in_range)
+{
+    if (unlikely(shutdown_called.load(std::memory_order_relaxed)))
+    {
+        const auto msg = fmt::format("Try to ingest files into a shutdown table, store={}", log->identifier());
+        LOG_FMT_WARNING(log, "{}", msg);
+        throw Exception(msg);
+    }
+
+    EventRecorder write_block_recorder(ProfileEvents::DMWriteFile, ProfileEvents::DMWriteFileNS);
+
+    auto delegate = dm_context->path_pool.getStableDiskDelegator();
+    auto file_provider = dm_context->db_context.getFileProvider();
+
+    size_t rows = 0;
+    size_t bytes = 0;
+    size_t bytes_on_disk = 0;
+
+    DMFiles files;
+    for (const auto & external_file : external_files)
+    {
+        auto file_parent_path = delegate.getDTFilePath(external_file.id);
+
+        // we always create a ref file to this DMFile with all meta info restored later, so here we just restore meta info to calculate its' memory and disk size
+        auto file = DMFile::restore(file_provider, external_file.id, external_file.id, file_parent_path, DMFile::ReadMetaMode::memoryAndDiskSize());
+        rows += file->getRows();
+        bytes += file->getBytes();
+        bytes_on_disk += file->getBytesOnDisk();
+
+        // Do some simple verification for the file range.
+        if (file->getRows() > 0)
+        {
+            RUNTIME_CHECK(external_file.range.has_value());
+            RUNTIME_CHECK(!external_file.range->none());
+        }
+        else
+        {
+            RUNTIME_CHECK(!external_file.range);
+        }
+
+        files.emplace_back(std::move(file));
+    }
+
+    bool ingest_using_split = false;
+    if (clear_data_in_range && bytes >= 4 * 1024 * 1024)
```

Comment:
maybe reuse `dm_context.delta_small_column_file_bytes` for the threshold is better